drivers/pcie/host/shell.c

```diff
@@ -13,6 +13,79 @@
 #include <zephyr/drivers/pcie/cap.h>
 #endif
 
+struct pci_cap_id_to_str {
+	uint32_t id;
+	char *str;
+};
+
+static struct pci_cap_id_to_str pci_cap_list[] = {
+	{ PCI_CAP_ID_PM,     "Power Management" },
+	{ PCI_CAP_ID_AGP,    "Accelerated Graphics Port" },
+	{ PCI_CAP_ID_VPD,    "Vital Product Data" },
+	{ PCI_CAP_ID_SLOTID, "Slot Identification" },
+	{ PCI_CAP_ID_MSI,    "Message Signalled Interrupts" },
+	{ PCI_CAP_ID_CHSWP,  "CompactPCI HotSwap" },
+	{ PCI_CAP_ID_PCIX,   "PCI-X" },
+	{ PCI_CAP_ID_HT,     "HyperTransport" },
+	{ PCI_CAP_ID_VNDR,   "Vendor-Specific" },
+	{ PCI_CAP_ID_DBG,    "Debug port" },
+	{ PCI_CAP_ID_CCRC,   "CompactPCI Central Resource Control" },
+	{ PCI_CAP_ID_SHPC,   "PCI Standard Hot-Plug Controller" },
+	{ PCI_CAP_ID_SSVID,  "Bridge subsystem vendor/device ID" },
+	{ PCI_CAP_ID_AGP3,   "AGP 8x" },
+	{ PCI_CAP_ID_SECDEV, "Secure Device" },
+	{ PCI_CAP_ID_EXP,    "PCI Express" },
+	{ PCI_CAP_ID_MSIX,   "MSI-X" },
+	{ PCI_CAP_ID_SATA,   "Serial ATA Data/Index Configuration" },
+	{ PCI_CAP_ID_AF,     "PCI Advanced Features" },
+	{ PCI_CAP_ID_EA,     "PCI Enhanced Allocation" },
+	{ PCI_CAP_ID_FPB,    "Flattening Portal Bridge" },
+	{ PCI_CAP_ID_NULL,   NULL },
+};
+
+static struct pci_cap_id_to_str pcie_cap_list[] = {
```

Comment:
Maybe convert these two arrays into functions with switch statements to return the strings (and NULL or "unknown" for out of range values)? If the standard adds more, we will be accessing out-of-band elements.

Reply:
good point

Reply:
wait (had to get coffee first): I only check for the capabilities we know already. I.e.: instead of going through the capabilities linked-list in PCIe config space, I loop over the array itself. It's heavier (pcie_get_cap/pcie_get_ext_cap will run through the linked list in config space each time they are called), but there is no other issue with unknown capabilities than being unable to list it.

Well, I could change the code to traverse the linked-list in config space once instead

Reply:
Do you mean PCI extended or PCI Express for `pcie` here? Bad word.

Reply:
Oh yeah... never mind. I was looking at it before my caffeine intake.